Fix integer literals parsing

agarciamontoro opened the pull request:
This PR fixes #834 with some changes:

- Modify `convertInt` so it returns the smallest representation allowed. See https://github.com/src-d/go-mysql-server/commit/acb2bc0f3bdae0ab853d3360bda24f0d7f11fea4.
- ~Add a rule to convert integer literals to the specified types in the schema in INSERT nodes. See https://github.com/src-d/go-mysql-server/commit/2eb11bbb08f3cc3c5b07247ba1c1e91032667e7e.~ After discussion with @erizocosmico (check https://github.com/src-d/go-mysql-server/pull/840#discussion_r333931390), I've completely removed the integer conversion rule and moved its logic to Insert.Execute function. Much less code, which is always better. See instead https://github.com/src-d/go-mysql-server/pull/840/commits/653fdde5b1d0d0a61c1931d187b1909b30571b44#diff-7e21a9fc108efea330a88d73d2a81023.
- Adapt several parts of the code that broke after the above changes, namely:
    - `getInt64Literal` function was modified so it converts any integer to int64. See https://github.com/src-d/go-mysql-server/commit/10d13fc633f6afc0a6a5214d3516cd2bc06beb74.
    - an explicit conversion to int64 was added in GROUP BY parsing, so any integer that fits is allowed. See https://github.com/src-d/go-mysql-server/commit/2320b073305611ddd8dc5e2a74291c1e00de5b22.
    - the ROUND and YEARWEEK functions were also modified so they can work with all types of integers. See https://github.com/src-d/go-mysql-server/commit/97a23158aa8444cf25171100c6baf7c6537124b9
    - most of the changes in this PR are in tests, as they were tailored to the previous int64-only design. These changes make the tests expect the right types. Review with care here, please :pray: : https://github.com/src-d/go-mysql-server/commit/00a883389ab70381cfd03341d5fb1b2913830285
    - some more boilerplate changes in pilosa code so it handles 8 and 16 bits signed and unsigned integers values as well. See https://github.com/src-d/go-mysql-server/commit/5c38a0ddcf90f5c197effdeb796439520839e33d
    - I also needed to modify the conversion back to `sqltypes.Value` in `numberT.SQL` function, as it failed when 8 and 16 bits signed and unsigned integers values were passed, resulting in empty rows when doing, for example, `SELECT 1`. See https://github.com/src-d/go-mysql-server/commit/ffae1d2959be5b6ae6edec0006e5045a8f01f277

Maybe the most comfortable way to review this is:
- Check first the new `convertInt` in https://github.com/src-d/go-mysql-server/commit/acb2bc0f3bdae0ab853d3360bda24f0d7f11fea4 (if you know a way of making that code a bit more beautiful, I'd be glad to hear it)
- Check then ~the new rule for integer conversion https://github.com/src-d/go-mysql-server/commit/2eb11bbb08f3cc3c5b07247ba1c1e91032667e7e and its test https://github.com/src-d/go-mysql-server/commit/22600f9252fb1a24b33b945c76952b2437466eed~ the modified `Insert.Execute` function: https://github.com/src-d/go-mysql-server/pull/840/commits/653fdde5b1d0d0a61c1931d187b1909b30571b44#diff-7e21a9fc108efea330a88d73d2a81023
- Review the minor modifications: https://github.com/src-d/go-mysql-server/commit/10d13fc633f6afc0a6a5214d3516cd2bc06beb74, https://github.com/src-d/go-mysql-server/commit/2320b073305611ddd8dc5e2a74291c1e00de5b22, https://github.com/src-d/go-mysql-server/commit/97a23158aa8444cf25171100c6baf7c6537124b9, https://github.com/src-d/go-mysql-server/commit/5c38a0ddcf90f5c197effdeb796439520839e33d and https://github.com/src-d/go-mysql-server/commit/ffae1d2959be5b6ae6edec0006e5045a8f01f277
- Review the changes in existing tests, which will be tedious: https://github.com/src-d/go-mysql-server/commit/00a883389ab70381cfd03341d5fb1b2913830285 and the new additions: https://github.com/src-d/go-mysql-server/pull/840/commits/03f9766afc6eca5baca5ee96ddaa24a29c2be292, https://github.com/src-d/go-mysql-server/pull/840/commits/2917915a7fd02c4e0ababb8021e7bb27c15c8f08, https://github.com/src-d/go-mysql-server/pull/840/commits/cc99d6c45e707413e5dcf73f876de20d30131032 and https://github.com/src-d/go-mysql-server/pull/840/commits/fe3a84c1343193eed1889c8b49b1aa39954d1aed.

This one is my first big contribution to go-mysql-server, so please review and test thoroughly. Comments and suggestions are more than welcome.

erizocosmico commented on sql/analyzer/resolve_insert_literals.go:
```diff
@@ -0,0 +1,82 @@
+  defer span.Finish()
+
+  return plan.TransformUp(originalNode, func(node sql.Node) (sql.Node, error) {
+
```
leftover blank line

agarciamontoro replied:
Done in https://github.com/src-d/go-mysql-server/pull/840/commits/6d4af9b32ba3d74f35ab0d518029332cb797ceac

erizocosmico commented on sql/parse/parse.go:
```diff
@@ -938,22 +944,51 @@ func isAggregateFunc(v *sqlparser.FuncExpr) bool {
+// base, to its smallest representation possible, out of:
+// int8, uint8, int16, uint16, int32, uint32, int64 and uint64
+func convertInt(value string, base int) (sql.Expression, error) {
+	i8, err := strconv.ParseInt(value, base, 8)
```
Instead of the great pyramid of Ghiza what do you think of doing this to keep it a little bit more flat? 😛 

```go
if i8, err := stconv.ParseInt(value, base, 8); err == nil {
    return expression.NewLiteral(int8(i8), sql.Int8), nil
}

if ui8, err := stconv.ParseUint(value, base, 8); err == nil {
    return expression.NewLiteral(unt8(ui8), sql.Uint8), nil
}

// the rest of the checks

ui64, err := strconv.ParseUint(value, base, 64)
if err != nil {
    return nil, err
}

return expression.NewLiteral(uint64(ui64), sql.Uint64), nil
```

agarciamontoro replied:
This is what I wanted to hear, that arrow shape was ugly. Thanks!

agarciamontoro replied:
Done in https://github.com/src-d/go-mysql-server/pull/840/commits/a7aa47f53e96d9088894e3f9efe03f6e2ac4ef83

erizocosmico commented on sql/analyzer/resolve_insert_literals.go:
```diff
@@ -0,0 +1,82 @@
+
+var errWrongNumberOfValues = errors.NewKind("the number of values to insert differ from the expected columns")
+
+func convertIntegerLiteralsInsert(ctx *sql.Context, analyzer *Analyzer, originalNode sql.Node) (sql.Node, error) {
```
I think we should let the table implementations handle this. When the literals are parsed, they are converted to the minimum type of int possible, then the table is in charge of converting it.

If we really want to make this transparent for table implementations, then I'd go for doing `column.Type.Convert(theLiteralValue)` inside the `InsertInto` execution method, but I would not add a new rule in the analyzer for this specific case.

WDYT @src-d/data-processing?

agarciamontoro replied:
I think that the conversion should be done by us. If we convert string literals to proper dates, we should do the same here. I'd argue that `int8` is as different from `uint64` as a date is different from a string.

But maybe it makes sense to do it inside InsertInto, not as a new rule. I'll wait for the rest of opinions.

agarciamontoro replied:
Solved in https://github.com/src-d/go-mysql-server/commit/653fdde5b1d0d0a61c1931d187b1909b30571b44.

erizocosmico commented on sql/parse/parse.go:
```diff
@@ -948,39 +948,35 @@ func isAggregateFunc(v *sqlparser.FuncExpr) bool {
-		}
-		return expression.NewLiteral(uint8(ui8), sql.Uint16), nil
+	if i8, err := strconv.ParseInt(value, base, 8); err == nil {
+    return expression.NewLiteral(int8(i8), sql.Int8), nil
```
i think your editor did not run gofmt to this part

agarciamontoro replied:
:thinking: It looks good on my side, I'll check

agarciamontoro replied:
Spaces instead of tabs, ofc. Force pushed as https://github.com/src-d/go-mysql-server/pull/840/commits/3034d2d03a9d2a485470e630873fa40b9d87f61e

agarciamontoro commented:
I'm just noticing that I did not run gofmt over the new files. I'll fix the formatting in the next commit.

agarciamontoro commented:
As per @erizocosmico's comment (check https://github.com/src-d/go-mysql-server/pull/840#discussion_r333931390), I've completely removed the integer conversion rule and moved its logic to Insert.Execute function. Much less code, which is always better.

I've added it as a new commit (https://github.com/src-d/go-mysql-server/pull/840/commits/653fdde5b1d0d0a61c1931d187b1909b30571b44) for if someone still prefers the other alternative, but I can rebase it before merging.

agarciamontoro commented:
The latest four commits add missing cases to tests of modified functions. I think that CodeCoverage should be happy now.

erizocosmico commented:
Can you rebase? there are some conflicts nowl

agarciamontoro commented:
Rebased. Codecov is still complaining, though. I'll take a look.

carlosms commented on go.mod:
```diff
@@ -26,3 +26,5 @@ require (
 	vitess.io/vitess v3.0.0-rc.3.0.20190602171040-12bfde34629c+incompatible
 )
+
+go 1.13
```
My understanding is that this allows the code to use go 1.13 features, but we are supposed to support the last 2 stable versions. Should we set this to 1.12 instead?

agarciamontoro replied:
Oops, my bad. I use go 1.13 by default and it changes the `go.mod` automatically, so I guess I did a `git add -a` and included it. I'll remove it. Thanks for the catch!

agarciamontoro commented:
Force pushed without changes in `go.mod`. Thanks for the catch, @carlosms!
